File: I - Vanilla TRS/C++ code/Vanilla TRS.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <random>
#include <math.h>
#include <any>
#include <iostream>

typedef std::vector<std::unordered_map<std::string, std::vector<double>>> typePaths;

//Market datas
auto S0 = 100.0;
auto sigma0 = 0.1;
auto r0 = 0.05;

//TRS datas
auto T = 1.0;
std::vector<double> TRSpaymentSchedule = { 0,25, 50, 75, 99 };
auto sTRS = 0.01;

//Model datas
auto kappa = 0.1;
auto theta = 0.1;
auto lamb = 0.1;
auto nu = 0.1;
auto etha = 0.1;
auto alpha = 0.02;
auto rhoSsigma = 0.1;
auto rhoSr = 0.1;
auto deltaT = 0.01;
auto lT = int(T / deltaT);
auto nbSimul = 100;

typePaths pathsGeneration(int nbSimul, int lT, double S0, double sigma0, double r0, double kappa, double theta, double lamb, double nu, double etha, double alpha, double rhoSsigma, double rhoSr)
{
    typePaths paths;
    for (auto i = 0; i != nbSimul; i++)
    {
        std::vector<double> phiS;
        std::vector<double> phitild;
        std::vector<double> phihat;
        for (auto k = 0; k != lT; k++)
        {
            std::default_random_engine generator;
            phiS.emplace_back(std::normal_distribution<double>{0, 1}(generator));
            phitild.emplace_back(std::normal_distribution<double>{0, 1}(generator));
            phihat.emplace_back(std::normal_distribution<double>{0, 1}(generator));
        }
        std::vector<double> S(lT);
        std::vector<double> sigma(lT);
        std::vector<double> r(lT);
        std::vector<double> intr(lT);
        S[0] = S0;
        sigma[0] = sigma0;
        r[0] = r0;
        intr[0] = r[0];
        for (auto k = 0; k != (lT - 1); k++)
        {
            r[k + 1] = (nu - etha * r[k]) * deltaT + sqrt(alpha * r[k]) * (rhoSr * sqrt(deltaT) * phiS[k] + sqrt(1 - rhoSr * rhoSr) * sqrt(deltaT) * phihat[k]) + r[k];
            sigma[k + 1] = kappa * (theta - sigma[k]) * deltaT + lamb * sqrt(sigma[k]) * (rhoSsigma * sqrt(deltaT) * phiS[k] + sqrt(1 - rhoSsigma * rhoSsigma) * sqrt(deltaT) * phitild[k]) + sigma[k];
            S[k + 1] = S[k] * (r[k] * deltaT + sqrt(sigma[k] * deltaT) * phiS[k]) + S[k];
            intr[k + 1] = intr[k] + r[k + 1];
        }
        typePaths interRes = { {"Asset",S},{"Volatility",sigma},{"Interest rate",r},{"Cum. interest rate",intr} };
        paths.emplace_back(interRes);
    }
    return paths;
}

std::vector<double> alphaBetaTRS(int TRSpaydate, int prevTRSpaydate, const typePaths& paths, int nbSimul)
{
    auto alphak = 0.0;
    auto betak = 0.0;
    for (auto i = 0; i != nbSimul; i++)
    {
        auto res1 = (paths[i].at("Asset")[TRSpaydate] - paths[i].at("Asset")[prevTRSpaydate]) / (exp(paths[i].at("Cum. interest rate")[TRSpaydate]));
        auto res2 = (paths[i].at("Asset")[prevTRSpaydate]) / (exp(paths[i].at("Cum. interest rate")[TRSpaydate]));
        alphak += res1;
        betak += res2;
    }
    alphak /= nbSimul;
    betak /= nbSimul;
    std::vector<double> res;
    res.emplace_back(alphak);
    res.emplace_back(betak);
    return res;
}

int getIndex(const std::vector<double>& v, int K)
{
    auto it = find(v.begin(), v.end(), K);
    if (it != v.end())
    {
        int index = it - v.begin();
        return index;
    }
    else 
    {
        return 1;
    }
}

std::vector<double> TRS(const std::vector<double>& TRSpaymentSchedule, double sTRS, const typePaths& paths, int nbSimul)
{
    auto alpha = 0.0;
    auto beta = 0.0;
    for (auto s = TRSpaymentSchedule.begin();s!= TRSpaymentSchedule.end()-1;s++)
    {
        auto payDate = getIndex(TRSpaymentSchedule, *s) + 1;
        auto prevPayDate = getIndex(TRSpaymentSchedule, *s);
        auto abRes = alphaBetaTRS(TRSpaymentSchedule[payDate], TRSpaymentSchedule[prevPayDate], paths, nbSimul);
        alpha += abRes[0];
        beta += abRes[1] * (TRSpaymentSchedule[payDate] - TRSpaymentSchedule[prevPayDate]);
    }
    auto resTemp = alpha - sTRS * beta;
    auto impliedRate = alpha / beta;
    std::vector<double> res;
    res.emplace_back(resTemp);
    res.emplace_back(impliedRate);
    return res;
}

int main()
{
    auto paths = pathsGeneration(nbSimul, lT, S0, sigma0, r0, kappa, theta, lamb, nu, etha, alpha, rhoSsigma, rhoSr);
    auto TRSres = TRS(TRSpaymentSchedule, sTRS, paths, nbSimul);
    std::cout << "TRS market value is" << TRSres[0] << std::endl;
    std::cout << "TRS implied rate is" << TRSres[1] << std::endl;
}

File: 1 - Vanilla TRS/Python code/Vanilla TRS.py
# -*- coding: utf-8 -*-
"""
Created on Sat Mar  4 13:55:04 2023

@author: moureaux pierre
"""

import numpy as np
from matplotlib import pyplot as plt

"Market datas"
S0 = 100
sigma0 = 0.1
r0 = 0.05

"TRS datas"
T = 1
TRSpaymentSchedule = (25,50,75,99)
sTRS = 0.01

"Model datas"
kappa = 0.1
theta = 0.1
lamb = 0.1
nu = 0.1
etha = 0.1
alpha = 0.02
rhoSsigma = 0.1
rhoSr = 0.1
deltaT = 0.01
lT = int(T/deltaT)
nbSimul = 1000

def pathsGeneration(nbSimul,lT,S0,sigma0,r0,kappa,theta,lamb,nu,etha,alpha,rhoSsigma,rhoSr):
    paths = np.zeros(nbSimul,dtype=dict)
    for i in range(nbSimul):
        phiS=np.random.normal(0, 1, lT)
        phitild=np.random.normal(0, 1, lT)
        phihat=np.random.normal(0, 1, lT)
        S = np.zeros(lT)
        sigma = np.zeros(lT)
        r = np.zeros(lT)
        intr = np.zeros(lT)
        S[0]=S0
        sigma[0]=sigma0
        r[0]=r0
        intr[0]=r[0]
        for k in range(lT-1):
            r[k+1] = (nu-etha*r[k])*deltaT + np.sqrt(alpha*r[k])*(rhoSr*np.sqrt(deltaT)*phiS[k]+np.sqrt(1-rhoSr*rhoSr)*np.sqrt(deltaT)*phihat[k])+r[k]
            sigma[k+1] = kappa*(theta-sigma[k])*deltaT + lamb*np.sqrt(sigma[k])*(rhoSsigma*np.sqrt(deltaT)*phiS[k]+np.sqrt(1-rhoSsigma*rhoSsigma)*np.sqrt(deltaT)*phitild[k])+sigma[k]
            S[k+1] = S[k]*(r[k]*deltaT+np.sqrt(sigma[k]*deltaT)*phiS[k]) + S[k]
            intr[k+1]=intr[k]+r[k+1]
        paths[i] = {"Asset": S, "Volatility": sigma, "Interest rate": r, "Cum. interest rate": intr}
    return paths

def alphaBetaTRS(TRSpaydate,prevTRSpaydate, paths, nbSimul):
    alphak = 0
    betak = 0
    for i in range(nbSimul):
        res1 = (paths[i]['Asset'][TRSpaydate]-paths[i]['Asset'][prevTRSpaydate])/(np.exp(paths[i]['Cum. interest rate'][TRSpaydate]))
        res2 = (paths[i]['Asset'][prevTRSpaydate])/(np.exp(paths[i]['Cum. interest rate'][TRSpaydate]))
        alphak += res1
        betak += res2
    alphak /= nbSimul
    betak /= nbSimul
    return [alphak,betak]

paths = pathsGeneration(nbSimul,lT,S0,sigma0,r0,kappa,theta,lamb,nu,etha,alpha,rhoSsigma,rhoSr)

def TRS(TRSpaymentSchedule, sTRS, paths,nbSimul):
    res = 0
    alpha = 0
    beta = 0
    for s in TRSpaymentSchedule:
        alpha += alphaBetaTRS(s,TRSpaymentSchedule[TRSpaymentSchedule.index(s)-1],paths,nbSimul)[0]
        beta += alphaBetaTRS(s,TRSpaymentSchedule[TRSpaymentSchedule.index(s)-1],paths,nbSimul)[1]*(s-TRSpaymentSchedule[TRSpaymentSchedule.index(s)-1])
    res = alpha - sTRS*beta
    impliedRate = alpha/beta
    return [res, impliedRate]

print("TRS market value is",TRS(TRSpaymentSchedule, sTRS, paths,nbSimul)[0])
print("TRS implied rate is",TRS(TRSpaymentSchedule, sTRS, paths,nbSimul)[1])




